File: src/onebrain-cli/src/cli/obp.rs
//! OBP transport projection only. No node, grants, discovery or retry owner here.
use clap::{Args, Subcommand};
use onebrain_node::vnext_product_runtime::obp::contract;
use reqwest::{Method, Url};
use serde_json::{json, Value};
use std::io::{BufRead, Read, Write};
use std::path::{Path, PathBuf};
use std::time::Duration;

const UNKNOWN: &str = "OBP outcome unknown; retain original key/payload/context, read status and reconcile before retry; no automatic replay";

#[derive(Args)]
pub(crate) struct ObpArgs {
    #[arg(long, global = true, default_value = "http://127.0.0.1:4280")]
    api_url: String,
    /// Prefer ONEBRAIN_API_TOKEN to avoid shell history disclosure.
    #[arg(long, global = true)]
    api_token: Option<String>,
    #[command(subcommand)]
    command: Command,
}

#[derive(Subcommand)]
enum Command {
    Status,
    SourceList(Files),
    ReservationList(Files),
    RouteStatus(Files),
    IntentStatus(Files),
    Configure(Files),
    /// Admit a host-owned input_ref; never upload source bytes or addresses.
    SourceAdmit(Files),
    /// Disable preserves replay floors; there is no source deletion operation.
    SourceSetEnabled(Files),
    Refresh(Files),
    /// Connect only to the exact expected_peer; sends no application payload.
    RouteRequest(Files),
    /// Schedule an eligible pending intent; never reset counters or consent.
    IntentRetry(Files),
    NetworkKill(Files),
    NetworkReenable(Files),
    /// Read recorded metadata only. local_not_found does not prove no effect.
    Reconcile {
        #[arg(long)]
        context_file: PathBuf,
        #[arg(long, value_parser = hex_id)]
        idempotency_key: String,
    },
}

#[derive(Args)]
struct Files {
    /// Exact saved status envelope; never automatically replace session fences.
    #[arg(long)]
    context_file: PathBuf,
    /// Exact accepted request DTO, including stable key/generation for mutations.
    #[arg(long)]
    payload_file: PathBuf,
}

fn hex_id(s: &str) -> Result<String, String> {
    contract::id(&json!(s))
        .map(|_| s.to_owned())
        .map_err(|_| "expected exactly 64 lowercase hexadecimal characters".into())
}
fn read_json(path: &Path) -> Result<Value, String> {
    let mut bytes = Vec::new();
    std::fs::File::open(path)
        .map_err(|_| "cannot open DTO/context file")?
        .take(contract::MAX_BYTES as u64 + 1)
        .read_to_end(&mut bytes)
        .map_err(|_| "cannot read DTO/context file")?;
    contract::parse(&bytes).map_err(|_| "invalid or over-budget JSON file".into())
}
fn validate_dto(name: &str, v: &Value) -> Result<(), String> {
    contract::validate(name, v)
        .map_err(|_| "invalid OBP DTO; reconcile original command before retry".into())
}
fn fields(v: &Value, required: &[&str], optional: &[&str]) -> Result<(), String> {
    contract::fields(v, required, optional).map_err(|_| UNKNOWN.into())
}
fn session(status: &Value) -> Result<Value, String> {
    validate_response("status", status, None)?;
    if status["data"]["available"] != true {
        return Err(
            "OBP service unavailable; host must explicitly supply the shared service and bindings"
                .into(),
        );
    }
    Ok(status["data"]["session"].clone())
}
fn operation(name: &str) -> Result<&'static Value, String> {
    contract::inventory()["operations"]
        .as_array()
        .unwrap()
        .iter()
        .find(|o| o["name"] == name)
        .ok_or_else(|| "unknown OBP operation".into())
}

struct Prepared {
    name: &'static str,
    route: &'static str,
    body: Option<Value>,
    management: bool,
}
impl Command {
    fn prepare(&self) -> Result<Prepared, String> {
        let (name, file) = match self {
            Self::Status => {
                return Ok(Prepared {
                    name: "status",
                    route: "status",
                    body: None,
                    management: false,
                })
            }
            Self::Reconcile {
                context_file,
                idempotency_key,
            } => {
                return Ok(Prepared {
                    name: "reconcile",
                    route: "reconcile",
                    management: false,
                    body: Some(
                        json!({"session":session(&read_json(context_file)?)?, "idempotency_key":idempotency_key}),
                    ),
                });
            }
            Self::SourceList(f) => ("source_list", f),
            Self::ReservationList(f) => ("reservation_list", f),
            Self::RouteStatus(f) => ("route_status", f),
            Self::IntentStatus(f) => ("intent_status", f),
            Self::Configure(f) => ("configure", f),
            Self::SourceAdmit(f) => ("source_admit", f),
            Self::SourceSetEnabled(f) => ("source_set_enabled", f),
            Self::Refresh(f) => ("refresh", f),
            Self::RouteRequest(f) => ("route_request", f),
            Self::IntentRetry(f) => ("intent_retry", f),
            Self::NetworkKill(f) => ("network_kill", f),
            Self::NetworkReenable(f) => ("network_reenable", f),
        };
        let op = operation(name)?;
        let payload = read_json(&file.payload_file)?;
        let command = op["effect"] != "none";
        let request = contract::Request::new(name, payload.clone(), command)
            .map_err(|_| "invalid OBP request DTO")?;
        Ok(Prepared {
            name,
            route: if command { "commands" } else { "query" },
            management: request.management(),
            body: Some(json!({
                "session":session(&read_json(&file.context_file)?)?, "operation":name, "payload":payload
            })),
        })
    }
}

struct Client {
    http: reqwest::Client,
    base: Url,
    token: String,
    management: Option<String>,
}
impl Client {
    fn new(args: &ObpArgs, management: Option<String>) -> Result<Self, String> {
        let base = Url::parse(&args.api_url).map_err(|_| "invalid local API origin")?;
        if !base
            .host_str()
            .unwrap_or("")
            .trim_matches(['[', ']'])
            .parse::<std::net::IpAddr>()
            .is_ok_and(|ip| ip.is_loopback())
            || !matches!(base.scheme(), "http" | "https")
            || base.path() != "/"
            || !base.username().is_empty()
            || base.password().is_some()
            || base.query().is_some()
            || base.fragment().is_some()
        {
            return Err("OBP requires a numeric loopback HTTP(S) origin without credentials, path, query or fragment".into());
        }
        let token = args
            .api_token
            .clone()
            .or_else(|| std::env::var("ONEBRAIN_API_TOKEN").ok())
            .filter(|s| !s.trim().is_empty())
            .ok_or("set ONEBRAIN_API_TOKEN or --api-token")?;
        let http = reqwest::Client::builder()
            .no_proxy()
            .redirect(reqwest::redirect::Policy::none())
            .retry(reqwest::retry::never())
            .connect_timeout(Duration::from_secs(5))
            .timeout(Duration::from_secs(30))
            .build()
            .map_err(|_| "cannot create local API client")?;
        Ok(Self {
            http,
            base,
            token,
            management,
        })
    }
    fn has_secret(&self, v: &Value) -> bool {
        contains_secret(v, &self.token)
            || self
                .management
                .as_ref()
                .is_some_and(|s| contains_secret(v, s))
    }
    async fn call(&self, p: &Prepared) -> Result<Value, String> {
        if p.management && self.management.as_ref().is_none_or(|s| s.is_empty()) {
            return Err(
                "host-scoped ONEBRAIN_OBP_MANAGEMENT_TOKEN required; CLI cannot mint authority"
                    .into(),
            );
        }
        let mut request = self
            .http
            .request(
                if p.body.is_some() {
                    Method::POST
                } else {
                    Method::GET
                },
                self.base
                    .join(&format!("/api/vnext/obp/{}", p.route))
                    .unwrap(),
            )
            .bearer_auth(&self.token);
        if p.management {
            request = request.header(
                "X-OneBrain-OBP-Management",
                self.management.as_ref().unwrap(),
            );
        }
        if let Some(body) = &p.body {
            let bytes = serde_json::to_vec(body).unwrap();
            if bytes.len() > contract::MAX_BYTES || self.has_secret(body) {
                return Err("request over budget or contains credentials in body".into());
            }
            request = request
                .header("Content-Type", "application/json")
                .body(bytes);
        }
        let mut response = request.send().await.map_err(|_| UNKNOWN)?;
        let status = response.status();
        let mut bytes = Vec::new();
        while let Some(chunk) = response.chunk().await.map_err(|_| UNKNOWN)? {
            if chunk.len() > contract::MAX_BYTES.saturating_sub(bytes.len()) {
                return Err(UNKNOWN.into());
            }
            bytes.extend_from_slice(&chunk);
        }
        let v = contract::parse(&bytes).map_err(|_| UNKNOWN)?;
        if self.has_secret(&v) {
            return Err(
                "API reflected credentials; reconcile original command before retry".into(),
            );
        }
        if !status.is_success() || v["ok"] != true {
            if matches!(status.as_u16(), 401 | 403) {
                return Err("OBP authentication/capability rejected; host must provide current scoped authority".into());
            }
            validate_error(&v)?;
            println!("{}", serde_json::to_string_pretty(&v).unwrap());
            return Err("OBP API failure; inspect error.obp; local_not_found is current-dataset only; no automatic replay".into());
        }
        validate_response(p.name, &v, p.body.as_ref())?;
        Ok(v)
    }
}
fn contains_secret(v: &Value, s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    match v {
        Value::String(v) => v.contains(s),
        Value::Array(a) => a.iter().any(|v| contains_secret(v, s)),
        Value::Object(m) => m
            .iter()
            .any(|(k, v)| k.contains(s) || contains_secret(v, s)),
        _ => false,
    }
}

fn envelope(v: &Value, ok: bool) -> Result<(), String> {
    fields(
        v,
        &["ok", "profile", if ok { "data" } else { "error" }, "meta"],
        &[],
    )?;
    if v["ok"] != ok || v["profile"] != onebrain_api::vnext_api::VNEXT_PRODUCT_PROFILE {
        return Err(UNKNOWN.into());
    }
    fields(
        &v["meta"],
        &["lifecycle", "coverage", "limitations", "continuation"],
        &[],
    )?;
    validate_dto("Lifecycle", &v["meta"]["lifecycle"])?;
    validate_dto("Coverage", &v["meta"]["coverage"])?;
    validate_dto("Limitations", &v["meta"]["limitations"])?;
    if !v["meta"]["continuation"].is_null() {
        validate_dto("Continuation", &v["meta"]["continuation"])?;
    }
    Ok(())
}
fn validate_error(v: &Value) -> Result<(), String> {
    envelope(v, false)?;
    fields(
        &v["error"],
        &["code", "message", "retryable", "limitations", "obp"],
        &[],
    )?;
    fields(
        &v["error"]["obp"],
        &["reason", "outcome", "reconcile_before_retry"],
        &["idempotency_key"],
    )?;
    let e = &v["error"]["obp"];
    let reason = e["reason"].as_str().ok_or(UNKNOWN)?;
    let (code, unknown, retryable) = match reason {
        "invalid_payload" => ("invalid_request", false, false),
        "local_not_found" => ("not_found", false, false),
        "generation_conflict" | "idempotency_conflict" | "session_conflict" => {
            ("conflict", false, false)
        }
        "input_expired" | "snapshot_expired" => ("expired", false, false),
        "admission_limit" => ("rate_limited", false, true),
        "feature_disabled" => ("capability_disabled", false, false),
        "dependency_unavailable" => ("dependency_unavailable", false, true),
        "outcome_unknown" => ("dependency_unavailable", true, true),
        "storage_corrupt" | "response_overflow" => ("internal_error", true, false),
        _ => return Err(UNKNOWN.into()),
    };
    if v["error"]["code"] != code
        || v["error"]["message"] != reason
        || v["error"]["retryable"] != retryable
        || e["reconcile_before_retry"] != unknown
        || e["outcome"] != if unknown { "unknown" } else { "not_admitted" }
    {
        return Err(UNKNOWN.into());
    }
    validate_dto("Limitations", &v["error"]["limitations"])?;
    if e.get("idempotency_key").is_some() {
        validate_dto("IdempotencyKey", &e["idempotency_key"])?;
    }
    Ok(())
}
fn validate_response(name: &str, v: &Value, request: Option<&Value>) -> Result<(), String> {
    envelope(v, true)?;
    let d = &v["data"];
    if name == "status" {
        if d["available"] == false {
            fields(d, &["available", "compiled"], &[])?;
            if !d["compiled"].is_boolean() || v["meta"]["lifecycle"] != "disabled" {
                return Err(UNKNOWN.into());
            }
        } else {
            fields(d, &["available", "session", "status"], &[])?;
            if d["available"] != true {
                return Err(UNKNOWN.into());
            }
            fields(
                &d["session"],
                &["process_generation", "dataset_generation"],
                &[],
            )?;
            for key in ["process_generation", "dataset_generation"] {
                validate_dto("IdempotencyKey", &d["session"][key])?;
            }
            validate_dto("ObpStatusV1", &d["status"])?;
        }
        return Ok(());
    }
    if name == "reconcile" || operation(name)?["effect"] != "none" {
        fields(
            d,
            &[
                "idempotency_key",
                "operation",
                "state",
                "reconcile_before_retry",
            ],
            if name == "reconcile" {
                &[]
            } else {
                &["result", "failure"]
            },
        )?;
        validate_dto("IdempotencyKey", &d["idempotency_key"])?;
        let op = operation(d["operation"].as_str().ok_or(UNKNOWN)?)?;
        if op["effect"] == "none" {
            return Err(UNKNOWN.into());
        }
        if name != "reconcile" && d["operation"] != name {
            return Err(UNKNOWN.into());
        }
        if let Some(r) = request {
            let key = if name == "reconcile" {
                &r["idempotency_key"]
            } else {
                &r["payload"]["idempotency_key"]
            };
            if &d["idempotency_key"] != key {
                return Err(UNKNOWN.into());
            }
        }
        let unknown = matches!(d["state"].as_str(), Some("admitted" | "reconcile_required"));
        if d["reconcile_before_retry"] != unknown {
            return Err(UNKNOWN.into());
        }
        match d["state"].as_str() {
            Some("completed") if name != "reconcile" => {
                if d.get("failure").is_some() {
                    return Err(UNKNOWN.into());
                }
                validate_dto(op["response"].as_str().ok_or(UNKNOWN)?, &d["result"])?;
                check_target(d["operation"].as_str().unwrap(), &d["result"], request)?;
            }
            Some("failed_no_effect") if name != "reconcile" => {
                if d.get("result").is_some() {
                    return Err(UNKNOWN.into());
                }
                let policy: Value = serde_json::from_str(include_str!(
                    "../../../test-vectors/vnext/obp-local-api-v1.json"
                ))
                .unwrap();
                if d["failure"]["outcome"] != "not_admitted"
                    || !policy["errors"].as_array().unwrap().contains(&d["failure"])
                {
                    return Err(UNKNOWN.into());
                }
            }
            Some("completed" | "failed_no_effect") if name == "reconcile" => {}
            Some("admitted" | "reconcile_required") => {
                if d.get("result").is_some() || d.get("failure").is_some() {
                    return Err(UNKNOWN.into());
                }
            }
            _ => return Err(UNKNOWN.into()),
        }
        return Ok(());
    }
    validate_dto(operation(name)?["response"].as_str().ok_or(UNKNOWN)?, d)?;
    check_target(name, d, request)
}
fn check_target(name: &str, data: &Value, request: Option<&Value>) -> Result<(), String> {
    if let Some(r) = request {
        let field = match name {
            "route_request" => "expected_peer",
            "route_status" => "route_id",
            "intent_status" | "intent_retry" => "intent_id",
            "source_set_enabled" => "source_id",
            _ => return Ok(()),
        };
        if data[field] != r["payload"][field] {
            return Err(UNKNOWN.into());
        }
    }
    Ok(())
}
fn confirmation(key: &str) -> Result<String, String> {
    eprint!("Type exact idempotency_key {key} to submit this bounded command: ");
    std::io::stderr()
        .flush()
        .map_err(|_| "cannot show confirmation")?;
    let mut typed = String::new();
    std::io::stdin()
        .lock()
        .take(128)
        .read_line(&mut typed)
        .map_err(|_| "cannot read confirmation")?;
    Ok(typed)
}
async fn run(
    args: &ObpArgs,
    management: Option<String>,
    confirm: impl FnOnce(&str) -> Result<String, String>,
) -> Result<Value, String> {
    let p = args.command.prepare()?;
    let client = Client::new(args, management)?;
    if p.management && client.management.as_ref().is_none_or(|s| s.is_empty()) {
        return Err(
            "host-scoped ONEBRAIN_OBP_MANAGEMENT_TOKEN required; CLI cannot mint authority".into(),
        );
    }
    if p.route == "commands" {
        let body = p.body.as_ref().unwrap();
        if client.has_secret(body) {
            return Err("credentials forbidden in command body".into());
        }
        eprintln!(
            "{}: {}",
            p.name,
            serde_json::to_string_pretty(&body["payload"]).unwrap()
        );
        let key = body["payload"]["idempotency_key"].as_str().unwrap();
        if confirm(key)?.trim_end_matches(['\r', '\n']) != key {
            return Err("command rejected: exact idempotency key confirmation required".into());
        }
    }
    client.call(&p).await
}
pub(crate) async fn execute(args: ObpArgs) -> Result<(), String> {
    let value = run(
        &args,
        std::env::var("ONEBRAIN_OBP_MANAGEMENT_TOKEN").ok(),
        confirmation,
    )
    .await?;
    println!("{}", serde_json::to_string_pretty(&value).unwrap());
    eprintln!("Bounded local view. Empty/path_limited is not global absence. Discovery and reservations confer no trust; command completion is not delivery, publication or reward. Preserve original keys and reconcile unknown outcomes before retry.");
    Ok(())
}
#[cfg(test)]
mod tests;

File: src/onebrain-web/src/api/obp.ts
import { fields, hexId, object, operations, parsePrivateJson, requireValue, validate } from './obpContract';
import type { Fields } from './obpContract';
import transportInventory from '../../../test-vectors/vnext/obp-local-api-v1.json';
export type Session = { process_generation: string; dataset_generation: string };
export type Meta = { lifecycle: string; coverage: string; limitations: string[]; continuation: null | string };
export type Status = Fields & { generation: number; limitations: string[] };
export type Snapshot = { available: false; compiled: boolean } | { available: true; session: Session; status: Status };
export type Result<T> = { data: T; meta: Meta };
export type Page = Fields & { items: Fields[]; continuation?: string; snapshot_frontier: string };
export type Recovery = { origin: string; session: Session; operation: string; payload: Fields };
export class ObpError extends Error {
  constructor(public reason: string, public uncertain = false) { super(reason); this.name = 'ObpError'; }
}
export function localOrigin(base: string): string {
  const u = new URL(base);
  requireValue(['http:', 'https:'].includes(u.protocol) && ['127.0.0.1', '[::1]'].includes(u.hostname) && !u.username && !u.password && !u.search && !u.hash && u.pathname === '/');
  return u.origin;
}
export function sessionValue(v: unknown): Session {
  const s = fields(v, ['process_generation', 'dataset_generation']);
  requireValue(hexId(s.process_generation) && hexId(s.dataset_generation)); return s as Session;
}
function metaValue(v: unknown): Meta {
  const m = fields(v, ['lifecycle', 'coverage', 'limitations', 'continuation']);
  validate('Lifecycle', m.lifecycle); validate('Coverage', m.coverage); validate('Limitations', m.limitations);
  if (m.continuation !== null) validate('Continuation', m.continuation); return m as Meta;
}
const errorRules = transportInventory.errors;
export function operationValue(v: unknown, operation: string, payload: Fields, metadata = false): Fields {
  const o = fields(v, ['idempotency_key', 'operation', 'state', 'reconcile_before_retry'], metadata ? [] : ['result', 'failure']);
  requireValue(o.idempotency_key === payload.idempotency_key && o.operation === operation);
  requireValue(['admitted', 'completed', 'failed_no_effect', 'reconcile_required'].includes(String(o.state)));
  const uncertain = o.state === 'admitted' || o.state === 'reconcile_required';
  requireValue(o.reconcile_before_retry === uncertain);
  if (metadata) return o;
  if (o.state === 'completed') {
    requireValue('result' in o && !('failure' in o));
    validate(operations.find(op => op.name === operation)!.response, o.result); matchTarget(operation, payload, object(o.result));
  } else if (o.state === 'failed_no_effect') {
    requireValue(!('result' in o)); const f = fields(o.failure, ['code', 'reason', 'retryable', 'outcome', 'reconcile_before_retry', 'http']);
    const rule = errorRules.find(r => r.reason === f.reason);
    requireValue(rule && rule.outcome === 'not_admitted' && Object.entries(rule).every(([k, v]) => f[k] === v));
  } else requireValue(!('result' in o) && !('failure' in o));
  return o;
}
function matchTarget(op: string, request: Fields, result: Fields) {
  const key = op === 'route_request' ? 'expected_peer' : op === 'route_status' ? 'route_id' : op.startsWith('intent_') ? 'intent_id' : op === 'source_set_enabled' ? 'source_id' : null;
  if (key) requireValue(request[key] === result[key]);
}
export function readRecovery(text: string): Recovery {
  const r = fields(parsePrivateJson(text), ['origin', 'session', 'operation', 'payload']);
  requireValue(typeof r.origin === 'string' && localOrigin(r.origin) === r.origin); sessionValue(r.session);
  const op = operations.find(o => o.name === r.operation && o.effect !== 'none'); requireValue(op); validate(op.request, r.payload);
  return r as Recovery;
}
export function createObpClient(connection: () => Promise<{ baseUrl: string; token: string }>, transport: typeof fetch = fetch) {
  let clientSession = '';
  async function request<T>(path: string, body?: unknown, management?: string): Promise<Result<T>> {
    const { baseUrl, token } = await connection(); const origin = localOrigin(baseUrl);
    const controller = new AbortController(); const timer = setTimeout(() => controller.abort(), 30000);
    try {
      const headers: Record<string, string> = { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' };
      if (management) headers['X-OneBrain-OBP-Management'] = management;
      if (clientSession) headers['X-OneBrain-VNext-Client-Session'] = clientSession;
      const response = await transport(`${origin}/api/vnext/obp/${path}`, { method: body ? 'POST' : 'GET', headers, body: body ? JSON.stringify(body) : undefined, cache: 'no-store', redirect: 'error', credentials: 'omit', referrerPolicy: 'no-referrer', signal: controller.signal });
      if (response.status === 401 || response.status === 403) throw new ObpError('Host authentication or scoped grant unavailable');
      requireValue(response.headers.get('content-type')?.toLowerCase().startsWith('application/json'));
      const reader = response.body?.getReader(); requireValue(reader); const chunks: Uint8Array[] = []; let length = 0;
      while (true) { const { done, value } = await reader.read(); if (done) break; length += value.byteLength; if (length > 1048576) { await reader.cancel(); throw new Error('Response bound'); } chunks.push(value); }
      const bytes = new Uint8Array(length); let offset = 0; for (const chunk of chunks) { bytes.set(chunk, offset); offset += chunk.length; }
      const envelope = object(parsePrivateJson(new TextDecoder('utf-8', { fatal: true }).decode(bytes)));
      requireValue(envelope.profile === 'VNEXT_PRODUCT_INTEGRATION_PROFILE_V1');
      const meta = metaValue(envelope.meta);
      if (envelope.ok === false) {
        const error = object(envelope.error);
        const detail = fields(error.obp, ['reason', 'outcome', 'reconcile_before_retry'], ['idempotency_key']);
        const rule = errorRules.find(r => r.reason === detail.reason);
        requireValue(rule && rule.http === response.status && rule.code === error.code && rule.retryable === error.retryable && rule.outcome === detail.outcome && rule.reconcile_before_retry === detail.reconcile_before_retry);
        throw new ObpError(String(detail.reason), rule.reconcile_before_retry);
      }
      requireValue(response.ok && envelope.ok === true); return { data: envelope.data as T, meta };
    } catch (e) { if (e instanceof ObpError) throw e; throw new ObpError('Transport or invalid response: read state and reconcile before retry', true); }
    finally { clearTimeout(timer); }
  }
  const client = {
    async origin() { return localOrigin((await connection()).baseUrl); },
    async status() {
      const r = await request<Snapshot>('status');
      try {
        if (r.data.available === false) { fields(r.data, ['available', 'compiled']); requireValue(typeof r.data.compiled === 'boolean'); }
        else { fields(r.data, ['available', 'session', 'status']); requireValue(r.data.available === true); sessionValue(r.data.session); validate('ObpStatusV1', r.data.status); }
      } catch { throw new ObpError('Invalid status projection', true); }
      return r;
    },
    async query<T = Fields>(session: Session, operation: string, payload: Fields): Promise<Result<T>> {
      sessionValue(session); const op = operations.find(o => o.name === operation && o.name !== 'status' && o.effect === 'none'); requireValue(op); validate(op.request, payload);
      const r = await request<T>('query', { session, operation, payload });
      validate(op.response, r.data); matchTarget(operation, payload, object(r.data)); return r;
    },
    async command(record: Recovery, management?: string) {
      requireValue(await client.origin() === record.origin); readRecovery(JSON.stringify(record));
      const op = operations.find(o => o.name === record.operation)!;
      if (op.access === 'host_management') requireValue(management && /^obm1\.[A-Za-z0-9_-]{43}$/.test(management));
      const r = await request<Fields>('commands', { session: record.session, operation: record.operation, payload: record.payload }, op.access === 'host_management' ? management : undefined);
      try { operationValue(r.data, record.operation, record.payload); } catch { throw new ObpError('Invalid command outcome; reconcile required', true); } return r;
    },
    async reconcile(record: Recovery) {
      requireValue(await client.origin() === record.origin);
      const fresh = await client.status();
      if (!fresh.data.available) throw new ObpError('Shared OBP service unavailable');
      if (fresh.data.session.dataset_generation !== record.session.dataset_generation) throw new ObpError('Dataset changed; original outcome remains unknown');
      const r = await request<Fields>('reconcile', { session: fresh.data.session, idempotency_key: record.payload.idempotency_key });
      operationValue(r.data, record.operation, record.payload, true); return r;
    },
    async subscribe(session: Session, onHint: () => void, onGap: () => void, socketFactory: (url: string) => WebSocket = url => new WebSocket(url)) {
      sessionValue(session);
      const r = await request<Fields>('ws/tickets', { session, subscriptions: ['network'] });
      fields(r.data, ['ticket', 'client_session', 'expires_at', 'session_expires_at', 'subscriptions', 'limitations']);
      requireValue(JSON.stringify(r.data.subscriptions) === '["network"]');
      validate('Count', r.data.expires_at); validate('Count', r.data.session_expires_at); validate('Limitations', r.data.limitations);
      requireValue(/^obw1\.[A-Za-z0-9_-]{43}$/.test(String(r.data.ticket)) && /^obw1\.[A-Za-z0-9_-]{43}$/.test(String(r.data.client_session)) && r.data.ticket !== r.data.client_session);
      const ws = socketFactory(`${(await client.origin()).replace(/^http/, 'ws')}/api/vnext/obp/ws?ticket=${encodeURIComponent(String(r.data.ticket))}`);
      let sequence = 0, closed = false;
      const stop = () => { if (closed) return; closed = true; clientSession = ''; ws.onclose = null; ws.onmessage = null; ws.onerror = null; ws.close(); };
      const gap = () => { if (closed) return; stop(); onGap(); };
      ws.onclose = gap; ws.onerror = gap;
      ws.onmessage = event => {
        try {
          requireValue(typeof event.data === 'string' && event.data.length <= 16384);
          const e = fields(parsePrivateJson(event.data), ['profile', 'event_type', 'sequence', 'timestamp', 'lifecycle', 'coverage', 'limitations', 'data']);
          requireValue(e.profile === 'OBP_PRIVATE_WEBSOCKET_PROFILE_V1' && e.sequence === sequence + 1);
          validate('Lifecycle', e.lifecycle); validate('Coverage', e.coverage); validate('Limitations', e.limitations); validate('Count', e.timestamp);
          if (sequence === 0) {
            requireValue(e.event_type === 'subscription_ready'); const d = fields(e.data, ['subscriptions', 'session_expires_at']);
            requireValue(JSON.stringify(d.subscriptions) === '["network"]'); validate('Count', d.session_expires_at); clientSession = String(r.data.client_session);
          } else {
            requireValue(e.event_type === 'network_state');
            const d = fields(e.data, ['compiled', 'requested', 'active', 'kill_switch', 'signer_ready', 'source_count', 'usable_reservations', 'authenticated_routes', 'pending_intents', 'advertisement_state', 'claims_global_completion', 'authorizes_reward']);
            validate('ObpStatusV1', { ...d, generation: 1, lifecycle: e.lifecycle, coverage: e.coverage, limitations: e.limitations });
            onHint();
          }
          sequence++;
        } catch { gap(); }
      };
      return stop;
    },
  };
  return client;
}
export type ObpClient = ReturnType<typeof createObpClient>;
